### PR TITLE
refactor(compiler): generate host listeners in DirectiveWrappers

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 1]
```diff
@@ -125,12 +149,12 @@ class DirectiveWrapperBuilder implements ClassBuilder {
     return createClassStmt({
       name: DirectiveWrapperCompiler.dirWrapperClassName(this.dirMeta.type),
       ctorParams: dirDepParamNames.map((paramName) => new o.FnParam(paramName, o.DYNAMIC_TYPE)),
-      builders: [{fields, ctorStmts}, this]
+      builders: [{fields, ctorStmts, methods: methods}, this]
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: remove `:methods`

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 2]
```diff
@@ -229,13 +256,84 @@ function addDetectChangesInHostPropsMethod(
...
+function addHandleEventMethod(hostListeners: BoundEventAst[], builder: DirectiveWrapperBuilder) {
+  const actionStmts: o.Statement[] = [];
+  const actionResultExprs: o.Expression[] = [];
+  hostListeners.forEach((hostListener, eventIdx) => {
+    const evalResult = convertActionBinding(
+        builder, null, o.THIS_EXPR.prop(CONTEXT_FIELD_NAME), hostListener.handler,
+        `sub_${eventIdx}`);
+    const trueStmts = evalResult.stmts;
+    if (evalResult.preventDefault) {
+      const preventDefaultVar = o.variable(`pd_${eventIdx}`);
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Add note: This actually contains whether to allow default.

Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Maybe: still add `.and(preventDefaultVar)` to reuse the same variable and don't rely on having exclusive event names.

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 3]
```diff
@@ -229,13 +256,84 @@ function addDetectChangesInHostPropsMethod(
...
+      actionStmts.push(preventDefaultVar.set(o.literal(true)).toDeclStmt(o.BOOL_TYPE));
+      actionResultExprs.push(preventDefaultVar);
+      trueStmts.push(preventDefaultVar.set(evalResult.preventDefault).toStmt());
+    }
+    actionStmts.push(
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Add TODO: convert to switch

File: modules/@angular/compiler/src/view_compiler/event_binder.ts [hunk 4]
```diff
@@ -6,183 +6,149 @@
...
-        (<o.Statement[]>[markPathToRootStart.callMethod('markPathToRootAsCheckOnce', []).toStmt()])
-            .concat(this._method.finish())
-            .concat([new o.ReturnStatement(resultExpr)]);
-    // private is fine here as no child view will reference the event handler...
-    this.compileElement.view.methods.push(new o.ClassMethod(
-        this._methodName, [this._eventParam], stmts, o.BOOL_TYPE, [o.StmtModifier.Private]));
-  }
+function collectEvents(
+    boundEvents: BoundEventAst[], directives: DirectiveAst[]): Map<string, EventSummary> {
+  let hasComponentHostListener = false;
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: remove

File: modules/@angular/compiler/src/view_compiler/event_binder.ts [hunk 5]
```diff
@@ -6,183 +6,149 @@
...
-    eventListeners: CompileEventListener[]) {
-  Object.keys(directiveAst.directive.outputs).forEach(observablePropName => {
-    const eventName = directiveAst.directive.outputs[observablePropName];
-
-    eventListeners.filter(listener => listener.eventName == eventName).forEach((listener) => {
-      listener.listenToDirective(directiveInstance, observablePropName);
+function generateHandleEventMethod(
+    boundEvents: BoundEventAst[], directives: DirectiveAst[], compileElement: CompileElement) {
+  let hasComponentHostListener = false;
+  directives.forEach((dirAst) => {
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: use `directives.some`...

File: modules/@angular/compiler/src/view_compiler/event_binder.ts [hunk 6]
```diff
@@ -6,183 +6,149 @@
...
     }
   });
-}
+  boundEvents.forEach((renderEvent, renderEventIdx) => {
+    const evalResult = convertActionBinding(
+        compileElement.view, compileElement.view, compileElement.view.componentContext,
+        renderEvent.handler, `sub_${renderEventIdx}`);
+    const trueStmts = evalResult.stmts;
+    if (evalResult.preventDefault) {
+      const preventDefaultVar = o.variable(`pd_${renderEventIdx}`);
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: Use the same variable, see above

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

File: modules/@angular/compiler/src/view_compiler/event_binder.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: Done

File: modules/@angular/compiler/src/view_compiler/event_binder.ts [hunk 7]
```diff
@@ -6,183 +6,149 @@
...
+      hasComponentHostListener ? compileElement.appElement.prop('componentView') : o.THIS_EXPR;
+  const handleEventStmts = new CompileMethod(compileElement.view);
+  handleEventStmts.resetDebugInfo(compileElement.nodeIndex, compileElement.sourceAst);
+  handleEventStmts.push(markPathToRootStart.callMethod('markPathToRootAsCheckOnce', []).toStmt());
+  const eventNameVar = o.variable('eventName');
+  const actionResultExprs: o.Expression[] = [];
+  directives.forEach((dirAst, dirIdx) => {
+    const dirWrapper = compileElement.directiveWrapperInstance.get(dirAst.directive.type.reference);
+    if (dirAst.hostEvents.length > 0) {
+      const preventDefaultVar = o.variable(`dir_pd_${dirIdx}`);
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: Done

File: modules/@angular/compiler/src/view_compiler/event_binder.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: Done

File: modules/@angular/core/src/linker/view_utils.ts [hunk 8]
```diff
@@ -14,6 +14,7 @@ import {isPresent, looseIdentical} from '../facade/lang';
 import {ViewEncapsulation} from '../metadata/view';
 import {RenderComponentType, RenderDebugInfo, Renderer, RootRenderer} from '../render/api';
 import {Sanitizer} from '../security';
+import {Type} from '../type';
```
Comment (tbosch) [inline modules/@angular/core/src/linker/view_utils.ts]: Done

**Comment by tbosch** [comment]
CI fails because of a browserstack flake

**Comment by tbosch** [comment]
Landed in master as https://github.com/angular/angular/commit/32feb8a5328221eedaece8f5212b0d9b65210a42

**Comment by sod** [comment]
@tbosch minor inconvenience with this PR. If any NgModule declares a component twice, the resulting `ngfactory.ts` file outputs its `Wrapper_...Component` as two identical copies.

```typescript
@NgModule({
    declarations: [
        // one component defined twice. Makes no sense, but currently works in 2.1.2, and fails with this PR
        MyComponent,
        MyComponent,
    ],
})
export class MyModule {}
```

Maybe print an error if a component was declared twice?

**Comment by tbosch** [comment]
@sod thanks. Just created https://github.com/angular/angular/pull/12602 to fix this.

**Pull request closed without merging**